File: app/src/main/java/mm/pndaza/tipitakamyanmar/repository/SuttaRepository.java
package mm.pndaza.tipitakamyanmar.repository;

import android.database.Cursor;
import android.database.sqlite.SQLiteOpenHelper;

import java.util.ArrayList;

import mm.pndaza.tipitakamyanmar.model.Sutta;
import mm.pndaza.tipitakamyanmar.utils.SQLBuilder;

public class SuttaRepository extends BaseRepository {
    private static final String SUTTA_TABLE = "sutta";
    private static final String BOOK_TABLE = "book";
    private static final String SUTTA_NAME_COLUMN = "name";
    private static final String BOOK_ID_COLUMN = "book_id";
    private static final String BOOK_NAME_COLUMN = "name";
    private static final String BOOK_NAME_AS_COLUMN = "book_name";
    private static final String PAGE_NUMBER_COLUMN = "page_number";

    public SuttaRepository(SQLiteOpenHelper dbHelper) {
        super(dbHelper);
    }

    public ArrayList<Sutta> getSuttas(String filterWord) {
        ArrayList<Sutta> suttaList = new ArrayList<>();
        if (filterWord == null || filterWord.isEmpty()) {
            return suttaList;
        }
        SQLBuilder sqlBuilder = new SQLBuilder().
                select(SUTTA_TABLE + "." + SUTTA_NAME_COLUMN,
                        BOOK_ID_COLUMN,
                        BOOK_TABLE + "." + BOOK_NAME_COLUMN + " as " + BOOK_NAME_AS_COLUMN,
                        PAGE_NUMBER_COLUMN).
                from(SUTTA_TABLE).
                innerJoin(BOOK_TABLE, SUTTA_TABLE + "." + BOOK_ID_COLUMN, BOOK_TABLE + "." + "id").
                where(SUTTA_TABLE + "." + SUTTA_NAME_COLUMN, " LIKE ", "%" + filterWord + "%");
        String sql = sqlBuilder.build();
        Cursor cursor = getReadableDatabase().rawQuery(sql, null);
        if (cursor != null) {
            if (cursor.moveToFirst()) {
                do {
                    String name = cursor.getString(cursor.getColumnIndexOrThrow(SUTTA_NAME_COLUMN));
                    String bookId = cursor.getString(cursor.getColumnIndexOrThrow(BOOK_ID_COLUMN));
                    String bookName = cursor.getString(cursor.getColumnIndexOrThrow(BOOK_NAME_AS_COLUMN));
                    int pageNumber = cursor.getInt(cursor.getColumnIndexOrThrow(PAGE_NUMBER_COLUMN));
                    suttaList.add(new Sutta(name, bookId, bookName, pageNumber));
                } while (cursor.moveToNext());
            }
            cursor.close();
        }
//        Log.d(TAG, "getSuttas: " + suttas.size());
        return suttaList;
    }
}

File: app/src/main/java/mm/pndaza/tipitakamyanmar/repository/SuttaRepository.kt
package mm.pndaza.tipitakamyanmar.repository

import android.database.sqlite.SQLiteOpenHelper
import mm.pndaza.tipitakamyanmar.database.DatabaseSchema.BookTable
import mm.pndaza.tipitakamyanmar.database.DatabaseSchema.SuttaTable
import mm.pndaza.tipitakamyanmar.model.Sutta

class SuttaRepository(dbHelper: SQLiteOpenHelper) : BaseRepository(dbHelper) {

    val bookName = "book_name"

    fun getSuttas(filterWord: String): List<Sutta> {
        if (filterWord.isEmpty()) {
            return emptyList()
        }

        val suttaList = mutableListOf<Sutta>()
        val sql = """
            SELECT 
                ${SuttaTable.table}.${SuttaTable.name},
                ${SuttaTable.bookId},
                ${BookTable.table}.${BookTable.name} AS $bookName,
                ${SuttaTable.pageNumber}
            FROM ${SuttaTable.table}
            INNER JOIN ${BookTable.table} 
                ON ${SuttaTable.table}.${SuttaTable.bookId} = ${BookTable.table}.${BookTable.id}
            WHERE ${SuttaTable.table}.${SuttaTable.name} LIKE ?
        """.trimIndent()

        val selectionArgs = arrayOf("%$filterWord%")

        readableDatabase.rawQuery(sql, selectionArgs).use { cursor ->
            if (cursor.moveToFirst()) {
                do {
                    val name = cursor.getString(cursor.getColumnIndexOrThrow(SuttaTable.name))
                    val bookId = cursor.getString(cursor.getColumnIndexOrThrow(SuttaTable.bookId))
                    val bookName = cursor.getString(cursor.getColumnIndexOrThrow(bookName))
                    val pageNumber =
                        cursor.getInt(cursor.getColumnIndexOrThrow(SuttaTable.pageNumber))
                    suttaList.add(Sutta(name, bookId, bookName, pageNumber))
                } while (cursor.moveToNext())
            }
        }
        return suttaList
    }
}